src/Util/JSON.php

```diff
@@ -18,13 +18,29 @@ final class JSON
      *
      * @param mixed $data The data to encode
      *
+     * @return mixed
+     *
      * @throws JsonException If the encoding failed
      */
-    public static function encode($data): string
+    public static function encode($data)
     {
-        $encodedData = json_encode($data, JSON_UNESCAPED_UNICODE);
+        $options = JSON_UNESCAPED_UNICODE;
+
+        if (\PHP_VERSION_ID >= 70200) {
+            $options |= JSON_INVALID_UTF8_SUBSTITUTE;
+        }
+
+        $encodedData = json_encode($data, $options);
 
-        if (JSON_ERROR_NONE !== json_last_error() || false === $encodedData) {
+        // This should never happen on PHP >= 7.2 as the substitution of invalid
+        // UTF-8 characters is done internally. On lower versions instead, we
+        // try to sanitize the data ourselves before retrying encoding. If it
+        // fails again we throw an exception as usual.
+        if (JSON_ERROR_UTF8 === json_last_error()) {
+            $encodedData = json_encode(self::sanitizeData($data, 256), $options);
```

Comment:
I would set a lower default limit as default depth: if we encounter a loop in the serialization it would make a mess!

Reply:
I took this limit from the XDebug default configuration for the maximum level of nested function calls. We can lower it of course, but I'm not sure it's worth it because regardless of the value the result is always the same: once the limit is reached the serialization stops.